### PR TITLE
Fully implement HarJels

### PR DESCRIPTION
**Pull request by Algebro7**
Several aspects of HarJels are not currently implemented, as described in #164. Implementing these exposed quite a few design problems in my opinion, particular around the behavior of armor and equipment. I've done my best at implementing these cleanly and without completely overhauling all of the code, and the code appears to work fine according to Maelwys and my testing.

One major issue was how armor was calculated--basically, armor was treated as a single object using a state pattern when patchwork was not being used, but uses multiple states belonging to the parent state when patchwork was being used. This caused me to have to completely redo armor BVs and always treat armor as patchwork for the purpose of defensive BV calculation, to be able to accommodate HarJel's affect on armor for a single location. 

There were no clean, easy methods to find locations of equipment, so I had to implement some and modify others that were bugged. Exclusions had to be refactored because they never considered where items were located or how many of them were equipped. I also had to add a bunch of methods dealing specifically with HarJel, which I'm not a big fan of. Several GUI components had to be refactored because they did not correctly update the display when different pieces of equipment were assigned.

TL;DR: This PR works, but I would welcome some feedback on the design since I had to work around a number of bugs and limitations to deal with HarJels. I am fine with going ahead and merging if you don't see anything you consider a major code smell.

### CODE REVIEW

File: sswlib/src/main/java/components/BipedLoadout.java [hunk 1]
```diff
@@ -1653,6 +1653,76 @@ public ArrayList FindIndexes( abPlaceable p ) {
...
+                locations.add(new LocationIndex(i, LocationIndex.MECH_LOC_LA));
+            }
+            if (RACrits[i].LookupName().equals(lookupName)) {
+                locations.add(new LocationIndex(i, LocationIndex.MECH_LOC_RA));
+            }
+        }
+        return locations;
+    }
+
+    public boolean HasHarjel() {
```
Comment (Algebro7) [inline sswlib/src/main/java/components/BipedLoadout.java:1704]: I kind of hate how these methods are called "HasHarjel" but only check for harjel 2 and 3. The issue was that harjel 1 isn't relevant for basically anything, whereas 2/3 apply different armor type modifiers. I just couldn't think of a name that covered both HarJel 2 and 3, and I didn't want to split into two functions.

File: sswlib/src/main/java/components/MechArmor.java [hunk 2]
```diff
@@ -2569,6 +2582,24 @@ public boolean AllowHarJel()
...
+    public double GetHarjelMod(int index) {
+        if (HasHarjel(index, Harjel2Locs)) {
+            return 1.1;
+        } else if (HasHarjel(index, Harjel3Locs)) {
+            return 1.2;
+        }
+        return 1.0;
+    }
+
+    private boolean HasHarjel (int index, List<LocationIndex> locs) {
```
Comment (Algebro7) [inline sswlib/src/main/java/components/MechArmor.java:2594]: This probably shouldn't be duplicated here, I'm trying to think if there was a reason I didn't just move this to `*Loadout.java` instead

File: sswlib/src/main/java/utilities/CostBVBreakdown.java [hunk 3]
```diff
@@ -146,30 +146,81 @@ public String Render() {
...
             } else {
-                retval += String.format( "    %1$-67s %2$,8.2f", "CT Armor Factor (" + ( CurMech.GetArmor().GetLocationArmor( LocationIndex.MECH_LOC_CT ) + ModArmor[LocationIndex.MECH_LOC_CT] * 10 + CurMech.GetArmor().GetLocationArmor( LocationIndex.MECH_LOC_CTR ) + ModArmor[LocationIndex.MECH_LOC_CTR] * 10 ) + ") * Armor Type Modifier (" + CurMech.GetArmor().GetCTArmorType().GetBVTypeMult() + ")", CurMech.GetArmor().GetCTDefensiveBV( ModArmor ) ) + NL;
+                line = "CT Armor Factor (" + ( CurMech.GetArmor().GetLocationArmor( LocationIndex.MECH_LOC_CT ) + ModArmor[LocationIndex.MECH_LOC_CT] * 10 + CurMech.GetArmor().GetLocationArmor( LocationIndex.MECH_LOC_CTR ) + ModArmor[LocationIndex.MECH_LOC_CTR] * 10 ) + ") * Armor Type Modifier (" + CurMech.GetArmor().GetCTArmorType().GetBVTypeMult() + ")";
+                if (CurMech.GetLoadout().HasHarjel(LocationIndex.MECH_LOC_CT)) {
+                    line += " * HarJel Modifier (" + CurMech.GetArmor().GetHarjelMod(LocationIndex.MECH_LOC_CT) + ")";
+                }
```
Comment (Algebro7) [inline sswlib/src/main/java/utilities/CostBVBreakdown.java:169]: This method needs refactored so badly

File: sswlib/src/main/java/components/BipedLoadout.java (hunk 1 above)
Comment (WEKarnesky) [reply sswlib/src/main/java/components/BipedLoadout.java:1704]: I would go off of what it is later used for.  I'm not at my dev system and can't find where this is consumed at the moment, but I assume it's for calculation or restriction purposes?

Comment (Algebro7) [reply sswlib/src/main/java/components/BipedLoadout.java:1704]: Yep, I'm looking at renaming it to "HasHarjelMod" now, since that more clearly indicates that it's checking for the presence of a harjel armor mod, likewise with the second version that checks by location.

Comment (WEKarnesky) [reply sswlib/src/main/java/components/BipedLoadout.java:1704]: That's a good change.

File: sswlib/src/main/java/utilities/CostBVBreakdown.java (hunk 3 above)
Comment (WEKarnesky) [reply sswlib/src/main/java/utilities/CostBVBreakdown.java:169]: it would probably benefit from refactoring, but my biggest complaint is more the long string.  Even simply breaking this up into multiple lines so you can more quickly see the execution path and how it builds the string would help.

In a similar vein, outputting the calls to variables, then building the string off those variables would look miles better than just a >300 character line.

**Comment by Algebro7** [comment]
Ok, I did some refactoring and I'm feeling much better about this. ae2c9d9 moves the harjel functions to the general `MechArmor` class, and I've added some methods to `Loadout` that make it much easier to find equipment (including harjels).

**Comment by Algebro7** [comment]
[SSW_0.7.4.1.zip](https://github.com/Solaris-Skunk-Werks/solarisskunkwerks/files/5246066/SSW_0.7.4.1.zip)

File: sswlib/src/main/java/utilities/CostBVBreakdown.java (hunk 3 above)
Comment (WEKarnesky) [reply sswlib/src/main/java/utilities/CostBVBreakdown.java:169]: Now I see what you mean with the whole function in context...

**Review by WEKarnesky — APPROVED** [review]
